### PR TITLE
HBHE cleanup and speedup

### PR DESCRIPTION
**Pull request by mariadalfonso**
This PR save 10% of the CPU speed in view of the Run3.
Changes are expected within numerical precision level in the E/ArrivalTime of recHits.
Pass clang formats checks.

Validation plots are here
[PR27726.pdf](http://dalfonso.web.cern.ch/dalfonso/HCAL/PR27726.pdf)

Main gain is in the updatePulseShape and updateCov functions of Mahi.
Below igprof report (test done on Run2018A_HLTPhysics run#316457 PU50 ).

ORIGINAL:
```
           13.4  .........      62.43 / 63.03        MahiFit::phase1Apply(HBHEChannelInfo const&, float&, float&, bool&, float&) const [45]
[46]       13.4      62.43       3.03 / 59.40      MahiFit::doFit(std::array<float, 3ul>&, int) const
            9.7  .........      45.45 / 45.45        MahiFit::minimize() const [47]
            1.7  .........       7.75 / 7.75         MahiFit::updatePulseShape(double, Eigen::Matrix<double, -1, 1, 0, 19, 1>&, Eigen::Matrix<double, -1, 1, 0, 19, 1>&, Eigen::Matrix<$
            1.3  .........       6.19 / 6.19         MahiFit::calculateArrivalTime() const [80]
            0.0  .........       0.00 / 0.89         _init [204]
            0.0  .........       0.00 / 0.12         _dl_runtime_resolve_xsave [757]

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
            9.7  .........      45.45 / 62.43        MahiFit::doFit(std::array<float, 3ul>&, int) const [46]
[47]        9.7      45.45       0.52 / 44.92      MahiFit::minimize() const
            4.2  .........      19.42 / 19.42        MahiFit::nnls() const [53]
            3.5  .........      16.26 / 16.26        MahiFit::updateCov() const [55]
            1.2  .........       5.54 / 5.54         MahiFit::onePulseMinimize() const [83]
            0.8  .........       3.63 / 3.63         MahiFit::calculateChiSq() const [94]
            0.0  .........       0.07 / 0.89         _init [204]
```

NEW

```
           11.0  .........      57.34 / 57.54        MahiFit::phase1Apply(HBHEChannelInfo const&, float&, float&, bool&, float&) const [45]
[46]       11.0      57.34       3.78 / 53.55      MahiFit::doFit(std::array<float, 3ul>&, int) const
            8.1  .........      42.62 / 42.62        MahiFit::minimize() const [47]
            1.1  .........       5.76 / 5.76         MahiFit::updatePulseShape(double, Eigen::Matrix<double, -1, 1, 0, 19, 1>&, Eigen::Matrix<double, -1, 1, 0, 19, 1>&, Eigen::Matrix<$
le, -1, -1, 0, 19, 19>&) const [78]
            1.0  .........       5.17 / 5.17         MahiFit::calculateArrivalTime() const [81]
            0.0  .........       0.00 / 0.79         _init [211]

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
            8.1  .........      42.62 / 57.34        MahiFit::doFit(std::array<float, 3ul>&, int) const [46]
[47]        8.1      42.62       0.32 / 42.29      MahiFit::minimize() const
            4.6  .........      24.09 / 24.09        MahiFit::nnls() const [48]
            2.1  .........      10.73 / 10.73        MahiFit::updateCov() const [60]
            0.9  .........       4.65 / 4.65         MahiFit::onePulseMinimize() const [85]
            0.5  .........       2.80 / 2.80         MahiFit::calculateChiSq() const [107]
            0.0  .........       0.02 / 0.79         _init [211]

```

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mariadalfonso** [comment]
@slava77 
Let me know if something is needed from my side for this PRreview

@VinInn FYI

**Comment by slava77** [comment]
>  Let me know if something is needed from my side for this PRreview

this is \~next on my list, I didn't look at the PR yet.

**Comment by slava77** [comment]
@cmsbuild please test
just to refresh the two week old tests

**Comment by mariadalfonso** [comment]
@slava77 
The validation plots are linked
http://dalfonso.web.cern.ch/dalfonso/HCAL/PR27726.pdf 
I noted the numerical precision level

**Review by slava77 — COMMENTED** [review]
sorry for the delay in the review.

File: RecoLocalCalo/HcalRecAlgos/src/PulseShapeFunctor.cc [hunk 1]
```diff
@@ -119,12 +119,57 @@ namespace FitterFuncs {
     return;
   }
 
+  void PulseShapeFunctor::funcShape(std::array<double, HcalConst::maxSamples> &ntmpbin,
```
[the remark this answers is not in the document]
Comment (mariadalfonso) [reply RecoLocalCalo/HcalRecAlgos/src/PulseShapeFunctor.cc]: @slava77 unified the funcShape

**Comment by abdoulline** [comment]
Just to add (for the record, better late than never...)  -
our usual single-pion test (for Run3 specifically) shows no any difference in 11_0_0_pre7 (with this PR included) vs 10_0_0_pre6 : 
 https://cms-cpt-software.web.cern.ch/cms-cpt-software/General/Validation/SVSuite/HCAL/calo_scan_single_pi/11_X/11_0_0_pre7_run3_vs_11_0_0_pre6_run3_SinglePi/

**Pull request merged**